---
title: Integrative Genomic Analyses for Identification and Prioritization of Long
  Non-Coding RNAs associated with Autism
author: "Brian Gudenas"
date: "November 11th, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = 'C:/Users/Brian/Google Drive/ASD_lncRNAs')
```
## Concise Description
Hundreds of Autism Spectrum Disorder (ASD) risk genes have been detected which converge on specific biological pathways, however in a majority of cases causal genetic factors are unknown. Could long non-coding RNAs (lncRNAs) be involved in the dysregulation of biological processes dysregulated in ASD?

### Main Objectives
1) Identify lncRNAs differentially expressed (DE lncRNAs) in the ASD cortex (SRP007483)
2) Build a brain developmental gene co-expression network 
3) Map DE lncRNAs and known ASD risk genes onto the network
4) Filter and prioritize DE lncRNAs based on co-expression to ASD genes


```{r tximport, eval=TRUE, cache=TRUE, message=FALSE, echo=TRUE}
#load libraries
library(GenomicFeatures)
library(tximport)
library(readr)
library(DESeq2)
library(biomaRt)
library(WGCNA)
library(dplyr)
library(stringr)
library(GenomicRanges)
library(genefilter)
library(gplots)
library(RColorBrewer)
options(stringsAsFactors=FALSE)

# Load in all Supplementary Input Data
load(file="./Data/All_Sup_Data.RData")

colnames(txi.salmon$counts)= rownames(DEG_samples)

# Differential Gene Expression Analysis
dds = DESeqDataSetFromTximport(txi.salmon, DEG_samples, ~Class)
dds_DE =DESeq(dds)
res05 =results(dds_DE, alpha = 0.05, contrast = c("Class","ASD","Con")) 

# pdf("./Figures/MA_plot.pdf")
plotMA(res05, main="MA plot of ASD/Con fold change", ylim=c(-4,4))
# dev.off()
res05= res05[!is.na(res05$padj), ]  
rownames(res05) = unlist(lapply(strsplit(rownames(res05), "\\."),"[[",1)) ## remove trailing decimals

#Use biomaRt to exract gene biotypes
mart = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
map = getBM(mart = mart, attributes=c("ensembl_gene_id","gene_biotype","start_position","end_position","chromosome_name", "external_gene_name"),filters = "ensembl_gene_id", values=rownames(res05))

sig_match = match(rownames(res05),  map$ensembl_gene_id)
res05$gene_name = map$external_gene_name[sig_match]
res05$biotype = map$gene_biotype[sig_match]
res05$chromosome = map$chromosome_name[sig_match]
res05$start_pos = map$start_position[sig_match]
res05$end_pos = map$end_position[sig_match]
res05$entrez = map$entrezgene[sig_match]

## create filter of all lncRNA biotypes as defined by Ensembl
lncRNA_filter = c("3prime_overlapping_ncrna", "antisense","antisense RNA", "lincRNA","ncrna host","processed_transcript", "sense_intronic" , "sense_overlapping")

lnc_test=c()
for (biotype in res05$biotype) {
    lnc_test =c(lnc_test, (sum(grepl(biotype, lncRNA_filter)) > 0 ) )
}
res05$lncRNA = lnc_test

tab = data.frame(logFC = res05$log2FoldChange, negLogPval = -log10(res05$padj))

# pdf("./Figures/Volcano.pdf")
par(mar = c(5, 4, 4, 4))
plot(tab, cex = 0.6, xlab=expression(Log[2]~fold~change), pch=1,
     ylab = expression(-Log[10]~pvalue), ylim = c(0,13))
title(main = "Differentially Expressed Genes in the ASD brain")
lfc = 1
pval = 0.05
signGenes = (abs(tab$logFC) >= lfc & tab$negLogPval > -log10(pval) & !res05$lncRNA)
points(tab[signGenes, ], pch = 1, cex = 0.8, col = "red3")
signLncRNAs = (abs(tab$logFC) >= lfc & tab$negLogPval > -log10(pval) & res05$lncRNA)
points(tab[signLncRNAs, ], pch = 1, cex = 0.8, col = "blue")
abline(h = -log10(pval), col = "magenta", lty = 2, lwd = 2)
abline(v = c(-lfc, lfc), col = "green2", lty = 2, lwd= 2)
mtext(paste("pval =", pval), side = 4, at = -log10(pval), cex = 0.8, line = 0.5, las = 1)
mtext(c(paste("-", lfc, "fold"), paste("+", lfc, "fold")), side = 3, at = c(-lfc, lfc),
      cex = 0.8, line = 0.5)
legend(3, 13, c("LncRNA","Non-lncRNA"), pch=c(1,1), col=c("blue","red3"))
# dev.off()

DEG =res05[res05$padj <= 0.05, ]
# check # of significant DE lncRNAs
table(DEG$lncRNA)
## check % of lncRNAs from each lncRNA biotype
table(DEG$biotype[DEG$lncRNA])/sum(DEG$lncRNA)*100
#write.csv(res05,"./Data/DEG_Results_alpha05.csv")

```
We have identified `r as.numeric(table(res05$lncRNA)[[2]])` differentially expressed lncRNAs (DE lncRNAs). Next, we will construct a weighted gene co-expression network using the BrainSpan Developmental transcriptome to infer the biological functions of these lncRNAs.

```{r WGCNA, cache=TRUE, message=FALSE, eval=TRUE}
score_match = match(SFARI_scores$Gene.Symbol, SFARI$Gene.Symbol)
SFARI_scores$entrez = SFARI$Entrez.GeneID[score_match]

# convert age based on diff units to a single continous variable -- Months Post-conception (conception ~10 months)
clinical$Months.Post.Conception = 0
for (i in 1:nrow(clinical)) {
    age = as.numeric(strsplit(as.character(clinical$age[i]), " ")[[1]][1])
    unit = strsplit(as.character(clinical$age[i]), " ")[[1]][2]
    
    if (unit == "pcw"){ age = age/4 #convert to months
        }else if (unit == "mos") { age = age+10 ## already in months but need to add 10 months for conception
        }else if (unit == "yrs") { age = (age*12)+10}## convert years to months then add 10 for conception
    clinical$Months.Post.Conception[i]=age
}

# create unique IDs for Expr and clinical samples -------------------------
colnames(Expr)<-str_c(clinical$structure_acronym, clinical$Months.Post.Conception, clinical$gender,clinical$donor_id, sep="_")
rownames(clinical) = colnames(Expr)

# filter samples to only retain regions within the neocortex by matching " cortex" and not cerebellar cortex which is in the cerebellum and not neocortex
cortical_regions = grepl(" cortex", clinical$structure_name) & clinical$structure_name != "cerebellar cortex"
# double-check brain regions included in analysis
table(clinical$structure_name[cortical_regions])
#Remove the  primary motor-sensory cortex bc has only 5 samples while the rest all have ~30
cortical_regions = grepl(" cortex", clinical$structure_name) & clinical$structure_name != "cerebellar cortex" & clinical$structure_name != "primary motor-sensory cortex (samples)"

## total number of samples included for downstream analysis
sum(cortical_regions)

# filter clinical and Expr by cortical_region -----------------------------
clinical = clinical[cortical_regions, ]
Expr = Expr[ ,cortical_regions] 
rownames(Expr) = Bspan_rows$ensembl_gene_id

Gene_filter = rowVars(Expr) > 0.5
datExpr0 = Expr[Gene_filter, ] ## variance filter to remove lowly variable genes
datExpr0 = t(datExpr0)
dim(datExpr0)

##--Create Genelist dataframe containing all gene info from (DEG and SFARI)
genelist = Bspan_rows[Gene_filter,]
DEG_map = match(genelist$ensembl_gene_id, rownames(DEG))

genelist$L2FC = DEG$log2FoldChange[DEG_map]
genelist$L2FC[is.na(genelist$L2FC)] = 0    ### Give 0 to any gene not Differentially expressed
## number of DEG genes not present in the filtered BrainSpan Dataset
nrow(DEG) - table(genelist$L2FC!=0)[[2]]

mart = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
map = getBM(mart = mart, attributes = c("ensembl_gene_id","band","gene_biotype","chromosome_name","start_position","end_position"),
            filters = "ensembl_gene_id", values =  genelist$ensembl_gene_id)

mart_map = match(genelist$ensembl_gene_id, map$ensembl_gene_id)

genelist$biotype = map$gene_biotype[mart_map]
genelist$chromosome = map$chromosome_name[mart_map]
genelist$start = map$start_position[mart_map]
genelist$end = map$end_position[mart_map]

ASD_match = match(genelist$entrez_id, SFARI_scores$entrez)
genelist$ASD_score = SFARI_scores$Score[ASD_match]

lnc_test=c()
for (biotype in genelist$biotype) {
    lnc_test =c(lnc_test, (sum(grepl(biotype, lncRNA_filter)) > 0))
}
genelist$lncRNA = lnc_test

LncRNA_Gr = makeGRangesFromDataFrame(DEG[DEG$lncRNA, colnames(DEG)=="chromosome" | colnames(DEG)=="start_pos" |  colnames(DEG)=="end_pos"], start.field = "start_pos", end.field = "end_pos" )
genelist_Gr = makeGRangesFromDataFrame(genelist[genelist$lncRNA != TRUE & !is.na(genelist$start), colnames(genelist)=="chromosome" | colnames(genelist)=="start" |  colnames(genelist)=="end" ], start.field = "start", end.field = "end" )
names(genelist_Gr) = genelist$gene_symbol[genelist$lncRNA != TRUE & !is.na(genelist$start) ]

near_genes = names(genelist_Gr)[nearest(x = LncRNA_Gr, subject = genelist_Gr)]
DEG$nearest[DEG$lncRNA] = near_genes
DEG$nearest_ASD_score[DEG$lncRNA] = genelist$ASD_score[match( near_genes, genelist$gene_symbol)]
DEG$Gene_symbol  = genelist$gene_symbol[match(rownames(DEG), genelist$ensembl_gene_id)]

table(rownames(clinical)==rownames(datExpr0)) ## verify clinical matches expression for all samples


gsg = goodSamplesGenes(datExpr0,verbose=4);
if (!gsg$allOK)
{
    if (sum(!gsg$goodGenes)>0)
        printFlush(paste("Removing genes:", paste(colnames(datExpr0)[!gsg$goodGenes], collapse = ", ")));
    if (sum(!gsg$goodSamples)>0)
        printFlush(paste("Removing samples:", paste(rownames(datExpr0)[!gsg$goodSamples], collapse = ", ")))
    datExpr0= datExpr0[gsg$goodSamples, gsg$goodGenes]
    genelist=genelist[gsg$goodGenes,]
}

datExpr0 <- log2(datExpr0+1)

powers = c(seq(8,14,by=1), seq(14,26, by=2));
rm(Expr, tab, txi.salmon, res05, dds, dds_DE, gsg, mart_map, near_genes, signLncRNAs, signGenes)

save.image("./Data/pre_network.RData")
```
Now we will check samples for outliers and subsequently create the gene co-expression network. It is important that this next code block is run on a machine with at least 16GB of RAM.
```{r create_network, eval=FALSE}
load("./Data/pre_network.RData")

sft=pickSoftThreshold(datExpr0, powerVector=powers, verbose=5, networkType="signed", corFnc = "bicor",corOptions = list(use = 'p', maxPOutliers = 0.1), blockSize = 20000)

sft
power=sft$fitIndices$Power[sft$fitIndices$SFT.R.sq > 0.80][1] ##select lowest power with R^2 > 0.8

net = blockwiseModules(datExpr0, power = power,
                       networkType="signed", minModuleSize = 50, maxBlockSize = 20000,
                       mergeCutHeight = 0.15, deepsplit=4, corType= "bicor",corOptions = list(use = 'p', maxPOutliers = 0.1),
                       numericLabels = TRUE, pamRespectsDendro = FALSE,
                       saveTOMs = TRUE,
                       saveTOMFileBase = "./Data/TOM",
                       verbose = 3 )
save.image("./Data/Network.RData")

```

lets check it out
```{r post-network, cache=TRUE}
load(file="./Data/Network.RData")
table(net$colors)

lncrnaColors = rep("grey", nrow(genelist))
DEGcolors = rep("grey", nrow(genelist))
ASDcolors = rep("grey", nrow(genelist))
lncrnaColors[genelist$lncRNA == TRUE] = "red3"
DEGcolors[genelist$L2FC != 0 ] = "blue3"
ASDcolors[!is.na(genelist$ASD_score) ] = "black"
moduleColors = labels2colors(net$colors)
genelist$Module = moduleColors

# pdf("./Figures/Dendrogram.pdf")
plotDendroAndColors(net$dendrograms[[1]], cbind(moduleColors,lncrnaColors, DEGcolors, ASDcolors),
                    c("Module","LncRNAs","DEG", "ASD"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
# dev.off()
moduleLabels = net$colors

MEs = net$MEs;
geneTree = net$dendrograms[[1]];
nGenes = ncol(datExpr0)
nSamples=nrow(datExpr0)

## calculate MEs with module colors
MEs0 <-moduleEigengenes(datExpr0,moduleColors)$eigengenes
MEs = orderMEs(MEs0)

counts = model.matrix( ~structure_acronym - 1, data = clinical)
colnames(counts) = str_replace(colnames(counts), pattern = "structure_acronym", replacement = "")
counts =counts[, colSums(counts)!=0]

clinical = cbind(clinical, counts)
clinical$Male = rep(0, nrow(clinical))
clinical$Male[clinical$gender=="M"] = 1
moduleTraitCor = cor(MEs, clinical[ ,c(8:ncol(clinical))], use = "p");
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);

textMatrix = paste(signif(moduleTraitCor, 2), "\n(",
                   signif(moduleTraitPvalue, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)

#pdf("./Figures/Module-trait_relationships.pdf",width=14,height=7)
labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = names(clinical[, c(8:ncol(clinical))]),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = greenWhiteRed(60),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = paste("Module-trait relationships") )
# dev.off()

age=as.data.frame(clinical$Months.Post.Conception)
names(age)="Months_Age"
modNames=substring(names(MEs),3)
geneModuleMembership = as.data.frame(cor(datExpr0, MEs, use = "p"));

genes<-colnames(datExpr0)
genes2annot <- match(genes, genelist$ensembl_gene_id)
sum(is.na(genes2annot))

MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");
geneTraitSignificance = as.data.frame(cor(datExpr0, age, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));

names(geneTraitSignificance) = paste("GS.", names(age), sep="");
names(GSPvalue) = paste("p.GS.", names(age), sep="");

geneInfo= data.frame(Ensembl_ID = genelist$ensembl_gene_id,
                     gene_symbol= genelist$gene_symbol,
                     Log2FC_DE = genelist$L2FC,
                     lncRNA=genelist$lncRNA,
                     ASD=genelist$ASD_score,
                     moduleColor = moduleColors,
                     geneTraitSignificance,
                     GSPvalue)
modOrder = order(-abs(cor(MEs, age, use = "p")));

for (mod in 1:ncol(geneModuleMembership))
{
    oldNames = names(geneInfo)
    geneInfo = data.frame(geneInfo, geneModuleMembership[, modOrder[mod]],
                          MMPvalue[, modOrder[mod]]);
    names(geneInfo) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep=""),
                        paste("p.MM.", modNames[modOrder[mod]], sep=""))
}
geneOrder = order(geneInfo$moduleColor, -abs(geneInfo$GS.Months_Age));
geneInfo = geneInfo[geneOrder, ]
write.csv(geneInfo, file = "./Data/geneInfo.csv",row.names = FALSE )

genelist$ASD_score = as.character(genelist$ASD_score)
genelist$ASD_score[is.na(genelist$ASD_score)]=0

moduleMembership = vector(mode = "numeric", length = nrow(geneInfo))
for (i in 1:nrow(geneInfo)){
    module = paste0("MM.",geneInfo$moduleColor[i])
    moduleMembership[i] = select(geneInfo,  match(module, colnames(geneInfo)))[i,]
}
geneInfo$Membership = moduleMembership
genelist$membership =geneInfo$Membership[match(genelist$ensembl_gene_id, geneInfo$Ensembl_ID)]
save.image("./Data/Post_geneinfo_network.RData")

my_palette <- colorRampPalette(c("dark green","green","white","white","red","dark red"))(n = 599)


Module_dist = cor(MEs0,MEs0)
Modules0= paste(paste0(toupper(substr(names(table(moduleColors)), 1, 1)), tolower(substring(names(table(moduleColors)), 2))))
rownames(Module_dist) = Modules0
colnames(Module_dist) = Modules0

#pdf("./Figures/ModuleCor_heatmap.pdf")
heatmap.2(Module_dist,trace="none",main="Module Correlation Matrix",RowSideColors=names(table(moduleColors)),notecol="black",key=TRUE,col=my_palette,symm=F,symkey=F,symbreaks=F)
#dev.off()
###### this code is for modular analysis of ASD genes and lncRNAS
genelist$Module = moduleColors

mod_sum<-matrix(data=NA,nrow=length(table(moduleColors)),ncol=4)
colnames(mod_sum) <- c("DE_lncRNAs","SFARI ASD Genes","ME16","total")
rownames(mod_sum) <- names(table(moduleColors))

ME16_match = match(genelist$ensembl_gene_id, ME16)
genelist$ME16=0
genelist$ME16[!is.na(ME16_match)]=1

LncRNAs_DE = as.list(table(factor(moduleColors[genelist$lncRNA== TRUE & genelist$L2FC != 0],lev=rownames(mod_sum))))
mod_sum[,1] = as.numeric(LncRNAs_DE)


SFARI = table(factor(moduleColors[genelist$ASD_score=="1S" | genelist$ASD_score=="2S" | genelist$ASD_score=="3S" | genelist$ASD_score=="4S" | genelist$ASD_score==1 | genelist$ASD_score==2  | genelist$ASD_score==3 | genelist$ASD_score==4 | genelist$ASD_score==5],lev=rownames(mod_sum)))
mod_sum[,2] = as.numeric(SFARI)

ME16 = table(factor(moduleColors[genelist$ME16==1],lev=rownames(mod_sum)))
mod_sum[,3] = as.numeric(ME16)

mod_totals <- as.list(table(moduleColors))
mod_sum[,4]= as.numeric(mod_totals)

mod_sum = mod_sum[mod_sum[,1] > 1, ] ## remove modules with no lncRNAs

total_genes = nGenes
mat_p = matrix(data=NA, nrow=nrow(mod_sum), ncol=3)
mat_or = matrix(data=NA, nrow=nrow(mod_sum), ncol=3)
for (row in 1:nrow(mod_sum)){
    for (col in 1:3) {
        mod_total <- as.numeric(mod_sum[row, 4])
        mod_count <- as.numeric(mod_sum[row, col])
        mod_non <- mod_total-mod_count
        non_count <- sum(as.numeric(mod_sum[,col])) - mod_count       
        non_non <- (total_genes-mod_total)-non_count
        
        contigency <- matrix(c(mod_count, mod_non, non_count, non_non),2,2)
        colnames(contigency) = c(rownames(mod_sum)[row], paste0("Non-", rownames(mod_sum)[row]))
        rownames(contigency) = c(colnames(mod_sum)[col], paste0("Non-", colnames(mod_sum)[col]))
        
        results <- fisher.test(contigency, alternative = "greater")
        p_val <- results[[1]]
        OR <- results[[3]]
        
        mat_p[row,col]<-p_val
        mat_or[row,col]<-OR
    }
}

rownames(mat_p)<-rownames(mod_sum)
rownames(mat_or)<-rownames(mod_sum)
colnames(mat_p)<-colnames(mod_sum)[1:3]
colnames(mat_or)<-colnames(mod_sum)[1:3]
### now correct P-values for each test using FDR method
 ad_p = p.adjust(mat_p, method = "fdr")
dim(ad_p) = dim(mat_p)
colnames(ad_p) <-colnames(mat_p)
rownames(ad_p) <- rownames(mat_p)
OR_filter<-matrix(data=NA,ncol=3,nrow=nrow(ad_p))
### * = p-value < 0.05 || ** = FDR-adjusted p-value < 0.05 ##### both need OR >= 1
OR_filter[ad_p <= 0.05] <- "*"
OR_filter[mat_or >= 1 & ad_p <= 0.05] <- paste0(round(mat_or[mat_or >= 1 & ad_p <= 0.05 ],2))
# log transform
ad_p<-log10(ad_p)*-1 
rownames(OR_filter)<-rownames(ad_p)

my_palette <- colorRampPalette(c("white","orange","red"))(n = 399)
 Modules_cap = paste0(toupper(substr(rownames(ad_p), 1, 1)), tolower(substring(rownames(ad_p), 2)))
rownames(ad_p) = Modules_cap

#pdf("./Figures/module_geneset_enrichment.pdf")
heatmap.2(ad_p, cellnote=OR_filter, trace="none", main="ASD Gene Set Enrichment by Module", RowSideColors=Modules_cap, notecol="black",key=TRUE, col=my_palette, symm=F, symkey=F, symbreaks=F, key.xlab="-Log( FDR adjusted p-value )", notecex =1.5, margins =c(13,8), breaks = seq(0, 15, length.out =400))

#dev.off()
```
Next we perform co-expression permutations to quantify the strength to which DE lncRNAs are co-expressed with ASD risk genes.
```{r co-expression permutation, message = FALSE, cache = TRUE}
lncRNAlist = genelist[genelist$lncRNA & genelist$L2FC !=0, ]

ASD_lncMatch = match( lncRNAlist$ensembl_gene_id , colnames(datExpr0))
genelist$ASD_score[is.na(genelist$ASD_score)] = 0
ASD_genes = genelist$ASD_score != 0 & genelist$ASD_score != "S" & genelist$ASD_score != "6" & genelist$ASD_score != "5" 

LncRNA_ASD_mat = bicor(datExpr0[ ,ASD_lncMatch], datExpr0[ ,ASD_genes], use="p", maxPOutliers = 0.1)
lncRNA_ASD_pairs=sum(abs(LncRNA_ASD_mat))

P=10000 ## iterations
sig_pairs=vector(mode = "numeric",length = P)
lnc_num = nrow(LncRNA_ASD_mat)
for (iter in 1:P){
    chosen = sample(1:ncol(datExpr0), lnc_num, replace = FALSE)
    chosen_ASD_mat = bicor( datExpr0[,chosen], datExpr0[ , ASD_genes], use="p", maxPOutliers = 0.1)
    sig_pairs[iter]= sum(abs(chosen_ASD_mat))
}
Z_scores = scale(c(sig_pairs,lncRNA_ASD_pairs)) ## calculate Z_scores of randomized distribution with actual value on end
Lnc_score = Z_scores[length(Z_scores)] ## Z_score of Actual lncRNA ASD pairs
pnorm(-abs(Lnc_score))

#pdf("./Figures/LncRNA_coexpression_ASD-risk-genes.pdf")
hist(sig_pairs,ylim=range(0,2600), xlab = "Summed Correlation", main = "LncRNAs co-expression to random gene-sets compared to ASD risk genes")
## Use color of line at LncRNAs real value to indicate significance
col="black"
if ( p.adjust(pnorm(-abs(Lnc_score)), n = 2, method = "fdr") <= 0.05 ) {
    col="red" }
abline(v=lncRNA_ASD_pairs, lwd = 3, col = col)

#dev.off()


####################
#Perform same permutation test for ME16 genes
ME16_genes = genelist$ME16==1

LncRNA_ASD_mat = bicor(datExpr0[,ASD_lncMatch], datExpr0[, ME16_genes], use="p", maxPOutliers = 0.1)

lncRNA_ASD_pairs=sum(abs(LncRNA_ASD_mat))
P=10000 ## iterations
sig_pairs=vector(mode = "numeric",length = P)
lnc_num = nrow(LncRNA_ASD_mat)
for (iter in 1:P){
    chosen = sample(1:ncol(datExpr0), lnc_num, replace = FALSE)
    chosen_ASD_mat = bicor( datExpr0[,chosen], datExpr0[,ME16_genes], use="p", maxPOutliers = 0.1)
    sig_pairs[iter]= sum(abs(chosen_ASD_mat))
}
Z_scores = scale(c(sig_pairs,lncRNA_ASD_pairs)) ## calculate Z_scores of randomized distribution with actual value on end
Lnc_score = Z_scores[length(Z_scores)] ## Z_score of Actual lncRNA ASD pairs
pval = pnorm(-abs(Lnc_score))


#pdf("./Figures/LncRNA_coexpression_ME16-risk-genes.pdf")
hist(sig_pairs, xlab = "Summed Correlation", main = "LncRNAs co-expression to random gene-sets compared to ME16 genes")
## Use color of line at LncRNAs real value to indicate significance
col="black"
if ( p.adjust(pnorm(-abs(Lnc_score)), n = 2, method = "fdr") <= 0.05 ) {
    col="red" }
abline(v=lncRNA_ASD_pairs, lwd = 3, col = col)
#dev.off()

DEG_match = match(rownames(DEG), genelist$ensembl_gene_id[ASD_genes])
DE_ASD = table(is.na(DEG_match))[1]
NotDE_ASD = length(genelist$ensembl_gene_id[ASD_genes]) - DE_ASD
DE_nonASD = nrow(DEG)-length(genelist$ensembl_gene_id[ASD_genes])
NotDE_nonASD = nrow(genelist) - nrow(DEG)

contingency = matrix(nrow=2,ncol=2, data = c(DE_ASD, DE_nonASD, NotDE_ASD, NotDE_nonASD))
rownames(contingency) = c("ASD","non-ASD")
colnames(contingency) = c("DE", "Not DE")
contingency
#         DE Not DE
# ASD       54    267
# non-ASD 1664  21035
fisher.test(contingency, "greater")

GTEx_samples = GTEx_samples[order(GTEx_samples$Tissue), ]  ## reorder samples to match GTEx columns (alphabetical)

match_GTEx = match(GTEx$Gene.Name, genelist$gene_symbol)
GTEx$module = genelist$Module[match_GTEx]
GTEx$lncRNA = genelist$lncRNA[match_GTEx]
GTEx$L2FC = genelist$L2FC[match_GTEx]

## remove genes not in network
GTEx = GTEx[!is.na(GTEx$module), ]


sub_GTEx = GTEx %>%
    group_by(module) %>%
    summarise(
        means = mean(Brain...Cortex)
    ) 
    ggplot(data = sub_GTEx, aes(module, means), color = module)+
    geom_bar( width=1, stat = "identity", aes(fill=module)) +
    scale_fill_identity()

exprMat = data.matrix(GTEx[, -c(1,2,56,57,58)])
rownames(exprMat) = GTEx$Gene.Name

##-- filter exprMat to remove tissues with less than 60 samples using GTEx_samples
exprMat = exprMat[GTEx$L2FC !=0 & GTEx$lncRNA , GTEx_samples$Number.of.RNASeq.Samples > 50]

brain_colors = rep("white", ncol(exprMat))
brain_colors[grepl("brain", str_to_lower(colnames(exprMat)))] = "blue"


exprMat = scale(t(exprMat))
exprMat = exprMat[ ,colSums(is.na(exprMat))!=nrow(exprMat)]

library(gplots)
library(RColorBrewer)
my_palette <- colorRampPalette(c("grey","white","orange","orangered","darkred","black"))(n = 599)
#pdf("./Figures/DE_lncRNAs_GTEX_heatmap.pdf", width = 16, height = 12)

heatmap.2(exprMat , dendrogram = "row", trace="none", main="LncRNA Expression by Tissue-Type", notecol="black", key=TRUE, col=my_palette, symm=F, symkey=F, symbreaks=F, key.xlab="Median FPKM Scaled by LncRNA",RowSideColors = brain_colors, margins = c(6,12))
#dev.off()

```

<!-- ```{r GO Enrichment, cache = TRUE, message = FALSE} -->
<!-- library(GOstats) -->
<!-- library(org.Hs.eg.db) -->
<!-- library(ggplot2) -->
<!-- library(gridExtra) -->


<!-- terms_full =data.frame() -->
<!-- for ( mod in unique(moduleColors)){ -->

<!-- paramMF <- new("GOHyperGParams", geneIds = genelist$entrez_id[genelist$Module== mod], universeGeneIds = genelist$entrez_id,  -->
<!--              ontology = c("MF"), annotation = "org.Hs.eg", pvalueCutoff = 0.01, testDirection = "over") -->

<!-- paramBP <- new("GOHyperGParams", geneIds = genelist$entrez_id[genelist$Module== mod], universeGeneIds = genelist$entrez_id,  -->
<!--              ontology = c("BP"), annotation = "org.Hs.eg", pvalueCutoff = 0.01, testDirection = "over") -->
<!-- hypMF <- hyperGTest(paramMF) -->
<!-- hypBP <- hyperGTest(paramBP) -->

<!-- termsMF = summary(hypMF) -->
<!-- termsMF$type = "MF" -->
<!-- termsMF$Module =  mod -->
<!-- termsMF = rename(termsMF, GO_ID = GOMFID) -->

<!-- termsBP = summary(hypBP) -->
<!-- termsBP$type = "BP" -->
<!-- termsBP$Module =  mod -->
<!-- termsBP = rename(termsBP, GO_ID = GOBPID) -->

<!-- termsMF$Pvalue = p.adjust(termsMF$Pvalue, method = "fdr" , n = length(termsMF$Pvalue)) -->
<!-- termsBP$Pvalue = p.adjust(termsBP$Pvalue, method = "fdr" , n = length(termsBP$Pvalue)) -->

<!-- terms = rbind(termsMF[1:3, ], termsBP[1:3, ]) -->

<!-- terms_full = rbind(terms_full, terms) -->
<!-- } -->

<!-- terms_full = terms_full[rowSums(is.na(terms_full)) != ncol(terms),  ] -->

<!-- write.csv(terms_full, "./WGCNA/Data/GOEnrichmentTable.csv") -->

<!-- dir.create("./WGCNA/Figures/Expression_Trajectory") -->


<!-- for (mod in 1:ncol(MEs0)){ -->
<!--     modName = names(table(moduleColors))[mod] -->

<!--     mod_terms = terms_full[terms_full$Module==modName, ] -->
<!--     mod_terms = mod_terms[order(mod_terms$Pvalue), ] -->
<!--     term_names = mod_terms$Term[order(mod_terms$Pvalue)] -->

<!--     p1 = ggplot(mod_terms, aes( y=-log10(Pvalue), x=reorder(Term, Pvalue), fill= type)) + -->
<!--         geom_bar( stat = "identity", position = position_dodge()) + theme_bw() + -->
<!--         #theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         geom_hline(yintercept =  2, col = "red", lwd=1)+ -->
<!--         coord_flip()+ -->
<!--         ylab(expression(-Log[10]~Pvalue)) + xlab( "GO Term") +ggtitle(paste(paste0(toupper(substr(modName, 1, 1)), tolower(substring(modName, 2))), "Functional Enrichment") )+  -->
<!--         scale_fill_manual(values=c("#56B4E9",  "#E69F00")) -->

<!--     df = as.data.frame(cbind(Months=clinical$Months.Post.Conception[clinical$Months.Post.Conception <= 22], Expression = MEs0[,mod][clinical$Months.Post.Conception <= 22]) ) -->
<!--     p2= ggplot(df, aes(x = Months, y=Expression), main = paste(paste0(toupper(substr(modName, 1, 1)), tolower(substring(modName, 2))), "Developmental Expression"), -->
<!--                xlab="Post-Conception Developmental Months", ylab = "Module Eigengene (PC1)") + -->
<!--         geom_point(shape=1, col ="grey") + theme_bw() +geom_jitter(alpha=0.7, shape=1)+ -->
<!--         geom_smooth(method=loess, size=1, col= modName, alpha=.5)+ -->
<!--         ggtitle(label = paste0(toupper(substr(modName, 1, 1)), tolower(substring(modName, 2))) ) + -->
<!--         geom_vline(xintercept =  10, col="blue", lwd=1) -->

<!--     pdf(paste0("./WGCNA/Figures/Expression_Trajectory/", modName, ".pdf"), width = 12, height= 12) -->
<!--     grid.arrange(p1, p2, ncol=2) -->

<!--     dev.off() -->
<!-- } -->
<!-- ``` -->
